PNG writer doesn't use PixelAspectRatio

I'm not actually sure if this is a bug or a limitation of the PNG file format, but a PixelAspectRatio other than 1.0 is lost when converting to png.

Kinda fixed. Metadata is propagated correctly when outputting PNG, but unfortunately no viewer I tried respects it (imagemagick's display, RV, nuke and ffmpeg). Though if I took the same png and converted it back to dpx, RV showed it with the correct PAR.

Is there a way to bake the PAR into the output with oiiotool? My use case is generating quicktimes with an OCIO color space baked in, so png is just an intermediate format.

BTW (unrelated), I had a build error (well, warning) until I commented out imagebufalgo_compare.cpp empty_sample_region:
/home/mvisser/workspace/oiio/src/libOpenImageIO/imagebufalgo_compare.cpp:648: error: ‘bool OpenImageIO::v1_6::empty_sample_region(const OpenImageIO::v1_6::ImageBuf&, OpenImageIO::v1_6::ROI)’ defined but not used

--- a/src/png.imageio/png_pvt.h
+++ b/src/png.imageio/png_pvt.h
@@ -502,13 +502,11 @@ write_info (png_structp& sp, png_infop& ip, int& color_type,
         spec.attribute ("DateTime", date);
     }
 
-    ImageIOParameter *unit=NULL, *xres=NULL, *yres=NULL;
-    if ((unit = spec.find_attribute("ResolutionUnit", TypeDesc::STRING)) &&
-        (xres = spec.find_attribute("XResolution", TypeDesc::FLOAT)) &&
-        (yres = spec.find_attribute("YResolution", TypeDesc::FLOAT))) {
-        const char *unitname = *(const char **)unit->data();
-        const float x = *(const float *)xres->data();
-        const float y = *(const float *)yres->data();
+    string_view unitname = spec.get_string_attribute ("ResolutionUnit");
+    float xres = spec.get_float_attribute ("XResolution");
+    float yres = spec.get_float_attribute ("YResolution");
+    float paspect = spec.get_float_attribute ("PixelAspectRatio");
+    if (xres || yres || paspect || unitname.size()) {
         int unittype = PNG_RESOLUTION_UNKNOWN;
         float scale = 1;
         if (Strutil::iequals (unitname, "meter") || Strutil::iequals (unitname, "m"))
@@ -520,8 +518,23 @@ write_info (png_structp& sp, png_infop& ip, int& color_type,
             unittype = PNG_RESOLUTION_METER;
             scale = 100.0/2.54;
         }
-        png_set_pHYs (sp, ip, (png_uint_32)(x*scale),
-                      (png_uint_32)(y*scale), unittype);
+        if (paspect) {
+            // If pixel aspect is given, allow resolution to be reset
+            if (xres)
+                yres = 0.0f;
+            else
+                xres = 0.0f;
+        }
+        if (xres == 0.0f && yres == 0.0f) {
+            xres = 100.0f;
+            yres = xres * (paspect ? paspect : 1.0f);
+        } else if (xres == 0.0f) {
+            xres = yres / (paspect ? paspect : 1.0f);
+        } else if (yres == 0.0f) {
+            yres = xres * (paspect ? paspect : 1.0f);
+        }
+        png_set_pHYs (sp, ip, (png_uint_32)(xres*scale),
+                      (png_uint_32)(yres*scale), unittype);
     }
 
     // Deal with all other params

